### PR TITLE
Add 'Extract Task spec' CodeAction

### PR DESCRIPTION
**Pull request by evidolob**
Fix: #511 

Demo:

https://user-images.githubusercontent.com/929743/113717015-d7ff8a80-96f3-11eb-88e7-91e80b51a2c5.mov

### CODE REVIEW

File: src/yaml-support/tkn-code-actions.ts [hunk 1]
```diff
@@ -97,6 +101,129 @@ class PipelineCodeActionProvider implements vscode.CodeActionProvider {
...
+    action.taskSpecEndPosition = endPos;
+    action.taskSpecText = document.getText(new vscode.Range(taskSpecStartPos, endPos));
+
+    return action;
+  }
+
+  private async resolveExtractTaskAction(action: ExtractTaskAction): Promise<vscode.CodeAction> {
+    return vscode.window.withProgress({location: vscode.ProgressLocation.Notification, cancellable: false, title: 'Extracting Task...' }, async (): Promise<vscode.CodeAction> => {
+      try {
+        const name = await vscode.window.showInputBox({ignoreFocusOut: false, prompt: 'Provide Task Name' });
```
Comment (sudhirverma) [inline src/yaml-support/tkn-code-actions.ts:167]: I think ignoreFocusOut should be true showInputBox and  showQuickPick  close when clicking anywhere else on screen

Comment (evidolob) [reply src/yaml-support/tkn-code-actions.ts:167]: Yes, good catch!

File: src/yaml-support/tkn-code-actions.ts [hunk 2]
```diff
@@ -97,6 +101,129 @@ class PipelineCodeActionProvider implements vscode.CodeActionProvider {
...
+    action.taskSpecText = document.getText(new vscode.Range(taskSpecStartPos, endPos));
+
+    return action;
+  }
+
+  private async resolveExtractTaskAction(action: ExtractTaskAction): Promise<vscode.CodeAction> {
+    return vscode.window.withProgress({location: vscode.ProgressLocation.Notification, cancellable: false, title: 'Extracting Task...' }, async (): Promise<vscode.CodeAction> => {
+      try {
+        const name = await vscode.window.showInputBox({ignoreFocusOut: true, prompt: 'Provide Task Name' });
+        const type = await vscode.window.showQuickPick(['Task', 'ClusterTask'], {placeHolder: 'Select Task Type:', canPickMany: false});
```
Comment (sudhirverma) [inline src/yaml-support/tkn-code-actions.ts:168]: You forgot to add `ignoreFocusOut: true` here

File: src/yaml-support/tkn-code-actions.ts [hunk 3]
```diff
@@ -97,6 +101,129 @@ class PipelineCodeActionProvider implements vscode.CodeActionProvider {
...
+        const virtDoc = this.getDocForExtractedTask(name, type, action.taskSpecText);
+
+        const saveError = await tektonVfsProvider.saveTektonDocument(virtDoc);
+        if (saveError) {
```
Comment (sudhirverma) [inline src/yaml-support/tkn-code-actions.ts:172]: I got error here `error converting YAML to JSON: yaml: line 4: did not find expected key`

virtDoc =
```
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: new-task
  {}

spec:
  resources:
  inputs:
    - name: pipeline-git
      type: git
  outputs:
    - name: pipeline-git
      type: git
steps:
  - args:
      - '-c'
      - |
        ls -al $(resources.inputs.pipeline-git.path)
    command:
      - bash
    image: ubuntu
    name: list
    resources: {}

File: src/yaml-support/tkn-code-actions.ts [hunk 4]
```diff
@@ -97,6 +101,129 @@ class PipelineCodeActionProvider implements vscode.CodeActionProvider {
...
+    action.taskSpecEndPosition = endPos;
+    action.taskSpecText = document.getText(new vscode.Range(taskSpecStartPos, endPos));
+
+    return action;
+  }
+
+  private async resolveExtractTaskAction(action: ExtractTaskAction): Promise<vscode.CodeAction> {
+    return vscode.window.withProgress({location: vscode.ProgressLocation.Notification, cancellable: false, title: 'Extracting Task...' }, async (): Promise<vscode.CodeAction> => {
+      try {
+        const name = await vscode.window.showInputBox({ignoreFocusOut: true, prompt: 'Provide Task Name' });
```
Comment (sudhirverma) [inline src/yaml-support/tkn-code-actions.ts:167]: showInputBox and showQuickPick should be shown before progress.

File: src/yaml-support/tkn-code-actions.ts (hunk 3 above)
Comment (evidolob) [reply src/yaml-support/tkn-code-actions.ts:175]: Now it should work

**Review by siamaksade — APPROVED** [review]

**Review by sudhirverma — APPROVED** [review]

**Pull request merged**